src/interpreter.rs

```diff
@@ -120,3 +129,1207 @@ bitflags::bitflags! {
+    //     follows the actual signature data.
+    let vch_sig_copy = vch_sig.clone();
+    // If the S value is above the order of the curve divided by two, its
+    // complement modulo the order could have been used instead, which is
+    // one byte shorter when encoded correctly.
+    // FIXME: This can return `false` without setting an error, which is not the expectation of the
+    //        caller.
+    Ok(PubKey::check_low_s(&vch_sig_copy))
+}
+
+fn is_defined_hashtype_signature(vch_sig: &ValType) -> bool {
+    if vch_sig.is_empty() {
+        return false;
+    };
+
+    HashType::from_bits(i32::from(vch_sig[vch_sig.len() - 1]), true).is_ok()
+}
+
+fn check_signature_encoding(
+    vch_sig: &Vec<u8>,
+    flags: VerificationFlags,
+) -> Result<bool, ScriptError> {
+    // Empty signature. Not strictly DER encoded, but allowed to provide a
+    // compact way to provide an invalid signature for use with CHECK(MULTI)SIG
+    if vch_sig.is_empty() {
+        return Ok(true);
+    };
+    if !is_valid_signature_encoding(vch_sig) {
+        return set_error(ScriptError::SigDER);
+    } else if flags.contains(VerificationFlags::LowS) && !is_low_der_signature(vch_sig)? {
+        // TODO: The C++ impl claims that `serror` is set here, but it isn’t, so this eventually
+        //       ends up as `UnknownError`, when it should probably be `SigHighS`.
+        return Ok(false);
+    } else if flags.contains(VerificationFlags::StrictEnc)
+        && !is_defined_hashtype_signature(vch_sig)
+    {
+        return set_error(ScriptError::SigHashType);
+    };
+    Ok(true)
+}
+
+fn check_pub_key_encoding(vch_sig: &ValType, flags: VerificationFlags) -> Result<(), ScriptError> {
+    if flags.contains(VerificationFlags::StrictEnc)
+        && !is_compressed_or_uncompressed_pub_key(vch_sig)
+    {
+        return Err(ScriptError::PubKeyType);
+    };
+    set_success(())
+}
+
+fn check_minimal_push(data: &ValType, opcode: PushValue) -> bool {
+    if data.is_empty() {
+        // Could have used OP_0.
+        return opcode == OP_0;
+    } else if data.len() == 1 && data[0] >= 1 && data[0] <= 16 {
+        // Could have used OP_1 .. OP_16.
+        return u8::from(opcode) == u8::from(OP_1) + (data[0] - 1);
+    } else if data.len() == 1 && data[0] == 0x81 {
+        // Could have used OP_1NEGATE.
+        return opcode == OP_1NEGATE;
+    } else if data.len() <= 75 {
+        // Could have used a direct push (opcode indicating number of bytes pushed + those bytes).
+        return usize::from(u8::from(opcode)) == data.len();
+    } else if data.len() <= 255 {
+        // Could have used OP_PUSHDATA.
+        return opcode == OP_PUSHDATA1;
+    } else if data.len() <= 65535 {
+        // Could have used OP_PUSHDATA2.
+        return opcode == OP_PUSHDATA2;
+    }
+    true
+}
+
+pub fn eval_script(
+    stack: &mut Stack<Vec<u8>>,
+    script: &Script,
+    flags: VerificationFlags,
+    checker: &dyn SignatureChecker,
```

Comment:
The C++ code takes a `consensusBranchId`. Is that intended to abstracted into the `SignatureChecker`? (It is only used for signature checking, I think.)

Reply:
Exactly, it was previously required because the library would compute the sighash, but now that's done by the caller via a callback.